crates/ruff_server/src/server/api/requests/code_action.rs

```diff
@@ -1,81 +1,222 @@
+}
+
+fn quick_fix<'d>(
+    edits: impl Iterator<Item = &'d DiagnosticEdit> + 'd,
+) -> impl Iterator<Item = CodeActionOrCommand> + 'd {
+    edits.map(|edit| {
+        let code = &edit.diagnostic_fix.code;
+        let title = edit
+            .diagnostic_fix
+            .kind
+            .suggestion
+            .clone()
+            .unwrap_or(edit.diagnostic_fix.kind.name.clone());
+        types::CodeActionOrCommand::CodeAction(types::CodeAction {
+            title: format!("{DIAGNOSTIC_NAME} ({code}): {title}"),
+            kind: Some(types::CodeActionKind::QUICKFIX),
+            edit: Some(types::WorkspaceEdit {
+                document_changes: Some(types::DocumentChanges::Edits(edit.document_edits.clone())),
+                ..Default::default()
+            }),
+            ..Default::default()
+        })
+    })
+}
+
+impl SupportedCodeAction {
+    fn kind(self) -> CodeActionKind {
+        match self {
+            Self::QuickFix => CodeActionKind::QUICKFIX,
+            Self::FixAll => CodeActionKind::SOURCE_FIX_ALL,
+            Self::OrganizeImports => CodeActionKind::SOURCE_ORGANIZE_IMPORTS,
+        }
+    }
+}
+
+fn fix_all<'d>(
+    edits: impl Iterator<Item = &'d DiagnosticEdit> + 'd,
+) -> impl Iterator<Item = CodeActionOrCommand> + 'd {
+    let edits_made: Vec<_> = edits
+        .filter(|edit| {
+            edit.diagnostic_fix
+                .fix
+                .applies(ruff_diagnostics::Applicability::Safe)
+        })
+        .collect();
+    let diagnostics_fixed = edits_made
+        .iter()
+        .map(|edit| edit.original_diagnostic.clone())
+        .collect();
+
+    (!edits_made.is_empty())
+        .then(move || {
+            edits_made
+                .into_iter()
+                .flat_map(|edit| edit.document_edits.iter())
+        })
+        .map(|changes| {
+            vec![
+                types::CodeActionOrCommand::CodeAction(types::CodeAction {
+                    title: format!("{DIAGNOSTIC_NAME}: Fix all auto-fixable problems"),
+                    diagnostics: Some(diagnostics_fixed),
+                    kind: Some(types::CodeActionKind::SOURCE_FIX_ALL),
                     edit: Some(types::WorkspaceEdit {
-                        document_changes: Some(types::DocumentChanges::Edits(changes)),
+                        document_changes: Some(types::DocumentChanges::Edits(
+                            changes.cloned().collect(),
+                        )),
                         ..Default::default()
                     }),
                     ..Default::default()
-                }))
-            })
-            .collect();
+                }), // TODO: implement command handler for the server
+                    /*
+                    types::CodeActionOrCommand::Command(types::Command {
+                        ...
+                    }
+                     */
+            ]
+        })
+        .into_iter()
+        .flatten()
+}
```

Comment:
Nice trick to have a single iterator type. 

Nit: I do struggle reading this code. A lot is going on, including flattening. I wonder if the complexity is worth it to avoid a single `Vec` allocation, considering that we end up allocating a vec on line 174 anyway (we could avoid that by using an array instead).

I do find the non iterator version easier to reason about (I believe that's the equivalent code but I find it hard to say because I don't fully understand the iterator version)

````
    let edits_made: Vec<_> = edits
        .filter(|edit| {
            edit.diagnostic_fix
                .fix
                .applies(ruff_diagnostics::Applicability::Safe)
        })
        .collect();

    if edits_made.is_empty() {
        return Vec::new();
    }

    let diagnostics_fixed = edits_made
        .iter()
        .map(|edit| edit.original_diagnostic.clone())
        .collect();

    vec![
        types::CodeActionOrCommand::CodeAction(types::CodeAction {
            title: format!("{DIAGNOSTIC_NAME}: Fix all auto-fixable problems"),
            diagnostics: Some(diagnostics_fixed),
            kind: Some(types::CodeActionKind::SOURCE_FIX_ALL),
            edit: Some(types::WorkspaceEdit {
                document_changes: Some(types::DocumentChanges::Edits(
                    edits_made
                        .into_iter()
                        .flat_map(|edit| edit.document_edits.iter())
                        .cloned()
                        .collect(),
                )),
                ..Default::default()
            }),
            ..Default::default()
        }), // TODO: implement command handler for the server
            /*
            types::CodeActionOrCommand::Command(types::Command {
                ...
            }
             */
    ]
```

Edit: I think for now it would even be sufficient to return a single `Option`.